### ISSUE TITLE
Logarithmic Depth Buffer

### ISSUE THREAD

**Issue by xeolabs** [xeokit/xeokit-sdk#254]
Sometimes it's necessary to use a **logarithmic depth buffer** if dealing with huge differences in scale in a single scene. 

## Benefits

1. Prevents distant objects from being clipped by the far clipping plane, since the plane can then be positioned very far away, 
2. prevents [Z-fighting](https://en.wikipedia.org/wiki/Z-fighting) between objects by improving the precision of WebGL depth testing for objects that are close to the viewpoint, and
3. improves the precision of Annotation occlusion culling.

#### Performance

Internally, xeokit's logarithmic depth buffer uses WebGL ````gl_FragDepth````, if available, which disables WebGL's [Early Fragment Test](https://www.khronos.org/opengl/wiki/Early_Fragment_Test) optimization and can cause a decrease in performance. 

#### SAO

Scalable Ambient Obscurance (SAO) doesn't work well with a very distant far clipping plane. Possibly SAO could be modified, to have it's own "virtual" far clipping plane, the distance of which can perhaps be derived from SAO ````scale````.

## Examples

* [PerformanceModel with batched double-precision geometries, log depth buf **enabled**](https://xeokit.github.io/xeokit-sdk/examples/logDepthBuf_PerformanceModel_batching_RTC.html)
* [PerformanceModel with batched double-precision geometries, log depth buf **disabled**](https://xeokit.github.io/xeokit-sdk/examples/logDepthBuf_PerformanceModel_batching_RTC_logDepthBufDisabled.html)
* [PerformanceModel with instanced double-precision geometries, log depth buf **enabled**](https://xeokit.github.io/xeokit-sdk/examples/logDepthBuf_PerformanceModel_instancing_RTC.html)
* [PerformanceModel with instanced double-precision geometries, log depth buf **disabled**](https://xeokit.github.io/xeokit-sdk/examples/logDepthBuf_PerformanceModel_instancing_RTC_logDepthBufDisabled.html.html)
* [Scene graph with double-precision geometries, log depth buf **enabled**](https://xeokit.github.io/xeokit-sdk/examples/logDepthBuf_SceneGraph_RTC.html)
* [Scene graph with double-precision geometries, log depth buf **disabled**](https://xeokit.github.io/xeokit-sdk/examples/logDepthBuf_SceneGraph_RTC_logDepthbufDisabled.html)
* [XKT model with double-precision geometries, log depth buf **enabled**](https://xeokit.github.io/xeokit-sdk/examples/logDepthBuf_XKT_Lyon.html)
* [XKT model with double-precision geometries, log depth buf **disabled**](https://xeokit.github.io/xeokit-sdk/examples/logDepthBuf_XKT_Lyon_logDepthBufDisabled.html)

## Usage

To detect if xeokit's logarithmic depth buffer is supported in the user's browser:

````javascript
const supported = Viewer.logarithmicDepthbufferSupported; // True if supported
````
If that returns ````true````, then we can configure our ````Viewer```` to have a logarithmic depth buffer:

````javascript
const viewer = new Viewer({
    canvasId: "myCanvas",
    logDepthBufferEnabled: true // <<--- Enable logarithmic depth buffer
});
````

A logarithmic depth buffer enables us to specify a very distant far clipping plane for our camera. 

We'll set the camera's near clipping plane to one micrometer, and the far clipping plane to to 100 billion light years away:

````javascript
viewer.camera.perspective.near = 1e-6;
viewer.camera.perspective.far = 1e27;

viewer.camera.ortho.near = 1e-6;
viewer.camera.ortho.far = 1e27;
````

We purposely allows users to set different clipping plane distances for each projection type.

### Disabled

The screenshot below shows a programmatically built ````PerformanceModel````, with clipping planes set as above, and logarithmic depth buffer disabled. Note the Z-fighting among the objects.

![Screenshot from 2021-01-07 20-35-25](https://user-images.githubusercontent.com/83100/104470289-d96c5b80-55b9-11eb-89dc-8368b20743fa.png)

### Enabled

The same scene, with logarithmic depth buffer enabled:

![Screenshot from 2021-01-07 20-37-06](https://user-images.githubusercontent.com/83100/104470313-df623c80-55b9-11eb-8138-6498cbf06801.png)

**Comment by xeolabs** [comment]
Can't release in 1.5.0 - needs a little more investigation on how to avoid low-precision depth artifacts when close to objects.

![Screenshot from 2020-11-09 18-32-27](https://user-images.githubusercontent.com/83100/98588192-921cd280-22cb-11eb-9f6a-fdbc5f769b18.png)

**Comment by xeolabs** [comment]
https://github.com/xeokit/xeokit-sdk/tree/logarithmic-depth-buffer

**Comment by Amoki** [comment]
```javascript
const supported = Viewer.logarithmicDepthbufferSupported; // True if supported
```

What feature is needed? I'd like to check the compatibility of browsers versions

**Comment by xeolabs** [comment]
This relies on support for the WebGL(1) extension ````[EXT_frag_depth](http://www.khronos.org/registry/webgl/extensions/EXT_frag_depth/)````.

Browsers can be checked for the presence of this extension [here](https://www.khronos.org/registry/webgl/sdk/tests/conformance/extensions/ext-frag-depth.html);

**Comment by Amoki** [comment]
![Capture d’écran de 2021-01-14 13-40-24](https://user-images.githubusercontent.com/1349751/104592234-2a3d8c00-566e-11eb-9ae3-3ede5ce194ed.png)
from: https://developer.mozilla.org/en-US/docs/Web/API/EXT_frag_depth#browser_compatibility

**Comment by xeolabs** [comment]
When we're running on a browser that lacks support, we should (at least):

- still set a very distant clipping plane, of say 100K, and
- ensure that AnnotationsPlugin has a surfaceOffset of at least ~0.3, so that Z-fighting doesn't mess with Annotation occlusion culling.

For some models, lack of log depth support is OK, and picking is still accurate. The Lyon model doesn't seem to require log depth buf at all: https://xeokit.github.io/xeokit-sdk/examples/#logDepthBuf_XKT_Lyon_logDepthBufDisabled

I expect log depth buf becomes mandatory only as soon as the 3D scene contains a mixture of extremely near and far objects.

**Comment by Amoki** [comment]
```javascript
const logarithmicDepthSupported = Viewer.logarithmicDepthBufferSupported;
const viewer = new Viewer({
  canvasId: this.engine3dCanvasId,
  transparent: true,
  spinnerElementId: this.spinnerCanvasId,
  enableOffsets: this.enableOffsets,
  logDepthBufferEnabled: logarithmicDepthSupported,
});
if (logarithmicDepthSupported) {
  viewer.camera.perspective.near = 1e-6;
  viewer.camera.perspective.far = 1e27;
  viewer.camera.ortho.near = 1e-6;
  viewer.camera.ortho.far = 1e27;
} else {
  viewer.camera.perspective.far = 100000.0;
  viewer.camera.ortho.far = 100000.0;
}
```

This is what I do.

But `logarithmicDepthBufferSupported` is not a static method so it is not accessible directly through the class and not an instance :/ (https://github.com/xeokit/xeokit-sdk/blob/master/src/viewer/Viewer.js#L146)

If we set it as a static method, we must update all calls from instances (like https://github.com/xeokit/xeokit-sdk/blob/568974e5b3e1002b0c6b81557431b92e3269f4e0/src/viewer/scene/PerformanceModel/lib/batching/occlusion/BatchingOcclusionShaderSource.js#L30)

**Comment by Amoki** [comment]
If you check `scene.viewer.logarithmicDepthBufferSupported` everytime, we can do:
```javscript
const viewer = new Viewer({
  canvasId: this.engine3dCanvasId,
  transparent: true,
  spinnerElementId: this.spinnerCanvasId,
  enableOffsets: this.enableOffsets,
  logDepthBufferEnabled: true,
});
const logarithmicDepthSupported = viewer.logarithmicDepthBufferSupported;
if (logarithmicDepthSupported) {
  viewer.camera.perspective.near = 1e-6;
  viewer.camera.perspective.far = 1e27;
  viewer.camera.ortho.near = 1e-6;
  viewer.camera.ortho.far = 1e27;
} else {
  viewer.camera.perspective.far = 100000.0;
  viewer.camera.ortho.far = 100000.0;
}
```

**Comment by xeolabs** [comment]
Oops. It should be static. And I was assuming that a JS static property would be always accessible as an instance property also. I'll fix.

**Comment by xeolabs** [comment]
Fixed in master - ````logarithmicDepthBufferSupported```` is now both a static and an instance property of ````Viewer````.

**Comment by Amoki** [comment]
![Capture d’écran de 2021-01-14 18-02-59](https://user-images.githubusercontent.com/1349751/104623673-e446ef00-5692-11eb-9f6a-cd929c38dcd8.png)

I have this with `logDepthBufferEnabled = false` AND `logDepthBufferEnabled = true` (with the 19_Marc_Antoine_Petit.xkt you already have)

**Comment by xeolabs** [comment]
For this model, it seems that the near plane causes inaccuracy when it's set below ````0.001````. At that value, I saw slight jittering, then as I decreased it to ````0.0001````, it looked like this:

![Screenshot from 2021-01-14 18-31-39](https://user-images.githubusercontent.com/83100/104627701-a9938580-5697-11eb-8668-da8a153f48da.png)

In any case, a general rule for depth buffer accuracy is to push the near plane back as far as possible. 

````1e-6```` is really just for fun, and is too close for BIM model viewing I think.

I think ````0.1```` is a reasonable value to use, or even as far as ````0.5````.
